### DISCUSSION TITLE
Interfacing ST7565 with PCF8574 I2C IO Expander

### DISCUSSION THREAD

**Discussion by Maes2ro** [olikraus/u8g2#1930] [General]
I am currently using a 128x64 LCD display with a ST7567 driver IC. I am operating in F_4W_SW_SPI and this works well. My controller is a Teensy 4.0.

I would like to use a PCF8574 IO expander to provide the connections via I2C instead of direct connections to the controller. I can see that you have I2C Constructors for some driver IC's, so I am wondering if it is possible to adapt these methods?

**Comment by olikraus** [comment]
The I2C interface is only supported for displays which support native I2C. Port expander are not supported by u8g2

**Reply by Maes2ro** [reply]
Thank you very much for that information. I have now redesigned my circuit to use software SPI.
I now have another question. 
The display I am working with (Midas MCCOG128064B12W-SPTLY) has an ST7565P dirver IC from Sitronix. Could you please suggest which of your 7565 constructors would be the most suitable?

[MCCOG128064B12W-SPTLY.pdf](https://github.com/olikraus/u8g2/files/9364460/MCCOG128064B12W-SPTLY.pdf)
[ST7565P_V2-3.pdf](https://github.com/olikraus/u8g2/files/9364464/ST7565P_V2-3.pdf)

**Comment by olikraus** [comment]
You need to test which driver fits best :(
It might be possible that you need to adjust contrast.

**Comment by Maes2ro** [comment]
What parameters change between versions of the driver? (I'm just thinking
of versions of the 7565 of course as I'm sure there are big differences
between chips)

**Comment by olikraus** [comment]
There are no versions, there are variants: The display controller might be the same, but the display itself might be different. Especially the wiring between the display controller and the display itself might differ. As a consequence there are multiple constructors for the same controller which reflect difference in wiring and type if the display itself.

Here is one small example: The controller ST7565 can drive a dot matrix display with 132 columns and 65 rows.
However your display has 128 columns and 64 rows. So obviously one row and four columns are hidden.
Is the hidden row on the top or the bottom?
Are the hidden columns on the left or right or even on both sides?
It all depends on the wiring between controller chip and the 128x64 LCD matrix.
Ideally it should be mentioned in the display datasheet how the 128x64 LCD is located inside the controller 132x64 matrix. However it is not even mentioned in the [MCCOG128064B12W-SPTLY.pdf](https://github.com/olikraus/u8g2/files/9364460/MCCOG128064B12W-SPTLY.pdf) datasheet from above.

Soo... you need to test this (actually you need to do some reverse engineering).

I suggest to use the FlipMode example. If all is ok, then there should be a solid frame next to the border all the time.

**Comment by Maes2ro** [comment]
Thank you very much for the detailed advice. That makes perfect sense.
At the moment I can't get any display at all, so I am considering two
problems. First, is the charge pump circuit working? And Second, is the
display being initialised?

I will follow your advice and look at the circuit details.

**Comment by Maes2ro** [comment]
It has been very interesting studying the data sheets and also looking at the voltages on my external circuit.
I am pretty sure now that the charge pump is not working, so the necessary voltages to drive the display are not being created. It's possible that I have built it incorreectly, but I am confident that the arrangement of capacitors matches those on both the display data sheet and the lower level ST7565 data. What I am measuring is all voltages the same on the capacitors so my thought is that the switches are not cycling.
The area that I am currently looking at are the commands to enable teh internal power supply and configure the charge pump. We can see these on page 28/52 of the ST7576P data. In particular the 'Power Control' and 'Set Booster' functions.
Do you have initialisation functions somewhere that set up these functions? I've been looking throught the header files, but it's quite a long task as I'm not clear on what I am looking for, so I thought I would start with asking the question.

**Comment by olikraus** [comment]
The internal charge pump is enabled by the init sequence. As a consequence you can't measure different voltages before sending this init sequence.

The init code is located in the u8x8_d files and includes the code to enable the charge pump:

https://github.com/olikraus/u8g2/blob/master/csrc/u8x8_d_st7565.c#L200

There might be multiple errors (not only charge pump related):
 * Is the wiring correct?
 * Did you use the correct u8g2 constructor and correct u8g2 constructor arguments (pin names)
 
Related the display and charge pump:
  * use 1uF ceramic caps for the charge pump
  * Additionally caps between VDD and VSS (pin 14 and pin 15): 100nF (ceramic) and 2.2 uF or larger (any type)

Display connection:
  * Did you apply 3.3V?
  * Did you use a signal level translation from 5V to 3.3V in case you have a 5V controller (e.g. Arduino Uno)
  * Do NOT connect backlight, this can be done later once everything works.


If you would sent pictures of your setup / schematic / code I would probably be able to help better.
